### PR TITLE
fix: leave slot filling to only dialog service using nlu (PL-1109)

### PR DESCRIPTION
**Pull request by theProf**
### Brief description. What is this change?

Remove slot filling from nlu service and leave solely to dialog service

### Implementation details. How do you make this change?

there are 2 places we do slot filling:
first in the nlu service
the in the dialog manager if needed
They both do the same fillSlots call (now, in my test). but the main difference is that the dialog manager adds a prefix to the utterance.
I'm unsure why the NLU performs poorly on the number match (only takes the first 3?) given the same intent and utterance (without prefix). but the solution i'll be pushing is to remove slot filling from the nlu service and leave all slot filling to be done in the dialog manager until a proper investigation can be done to tweak/optimize/etc

### Checklist

- [x] Appropriate tests have been written
- [x] bors try
- [ ] manual

### CODE REVIEW

**Comment by theProf** [comment]
bors try

File: lib/services/classification/predictor.class.ts [hunk 1]
```diff
@@ -314,6 +312,7 @@ export class Predictor extends EventEmitter {
     const nlcPrediction = !isIntentClassificationLLMSettings(this.settings) ? await this.nlc(utterance, false) : null;
     if (nlcPrediction) {
       this.predictions.result = 'nlc';
+      this.debug('NLC' as DebugType);
```
Comment (trs) [inline lib/services/classification/predictor.class.ts:315]: We don't want to expose NLC to the end user

Comment (theProf) [reply lib/services/classification/predictor.class.ts:315]: whoops, missed a debug line. thanks

File: lib/services/dialog/index.ts [hunk 2]
```diff
@@ -177,14 +176,30 @@ class DialogManagement extends AbstractManager<{ utils: typeof utils }> implemen
...
+            },
             {
               locale: version.prototype?.data.locales[0] as VoiceflowConstants.Locale,
               hasChannelIntents: project?.platformData?.hasChannelIntents,
               platform: version.prototype.platform as VoiceflowConstants.PlatformType,
             }
           );
-          dmPrefixedResult = getIntentRequest(await predictor.predict(`${prefix} ${query}`));
+
+          const filledSlots = await predictor.fillSlots(`${prefix} ${query}`, {
```
Comment (trs) [inline lib/services/dialog/index.ts:190]: Move this back to the `predict` call.

File: lib/services/dialog/index.ts [hunk 3]
```diff
@@ -177,14 +176,30 @@ class DialogManagement extends AbstractManager<{ utils: typeof utils }> implemen
               dmRequest: dmStateStore.intentRequest.payload,
               isTrained,
             },
-            intentClassificationSettings,
+            {
+              type: 'nlu',
+              params: { confidence: 0.6 },
```
Comment (trs) [inline lib/services/dialog/index.ts:181]: Is `version.platformData.settings.intentConfidence` still valid to use here?

**Review by trs — CHANGES_REQUESTED** [review]

File: lib/services/dialog/index.ts (hunk 3 above)
Comment (theProf) [reply lib/services/dialog/index.ts:181]: i wouldn't trust it as it's not a feature we currently explicitly support (storing both nlu + hybrid settings to switch between), you currently are expected to lose changes. Default (as in your PR) seems most solid right now

File: lib/services/classification/predictor.class.ts [hunk 4]
```diff
@@ -317,7 +296,7 @@ export class Predictor extends EventEmitter {
       return nlcPrediction;
     }
 
-    const nluPrediction = this.props.isTrained && (await this.nlu(utterance, this.options));
+    const nluPrediction = this.props.isTrained && (await this.nlu(utterance, { limit: 10 }));
```
Comment (theProf) [inline lib/services/classification/predictor.class.ts:299]: this should switch back to `this.options` with a default somewhere

File: lib/services/dialog/index.ts [hunk 5]
```diff
@@ -155,9 +158,15 @@ class DialogManagement extends AbstractManager<{ utils: typeof utils }> implemen
 
       try {
         const prefix = dmPrefix(dmStateStore.intentRequest.payload.intent.name);
-        const { intentClassificationSettings, intents, isTrained, slots } = castToDTO(version, project);
+        const { intents, isTrained, slots } = castToDTO(version, project);
+
+        const scopedIntent = intents?.find((intent) => intent.name === dmStateStore.intentRequest!.payload.intent.name);
+        const scopedSlots = scopedIntent?.slots?.map((slot) => slots?.find((entity) => entity.key === slot.id));
```
Comment (theProf) [inline lib/services/dialog/index.ts:164]:
```suggestion
        const scopedSlots = scopedIntent?.slots?.map((slot) => slots?.find((entity) => entity.key === slot.id)).filter(Boolean);
```
then below you don't have to `slot?.name`

File: lib/services/dialog/index.ts [hunk 6]
```diff
@@ -177,22 +186,42 @@ class DialogManagement extends AbstractManager<{ utils: typeof utils }> implemen
...
           );
-          dmPrefixedResult = getIntentRequest(await predictor.predict(`${prefix} ${query}`));
+
+          const prediction = await predictor.predict(`${prefix} ${query}`);
+
+          (prediction?.predictedSlots ?? []).forEach((predictedSlot) => {
+            const slot = scopedSlots?.find((scopedSlot) => scopedSlot?.name === predictedSlot.name);
+            if (!slot) return;
+
+            const existingEntity = slotFillingRequest.payload.entities.find((entity) => entity.name === slot.name);
```
Comment (theProf) [inline lib/services/dialog/index.ts:207]: pending size we may want to compute this once into a Map

File: lib/services/dialog/index.ts (hunk 5 above)
Comment (trs) [reply lib/services/dialog/index.ts:164]: Typescript isn't smart enough to know that slot is not undefined here

File: lib/services/dialog/index.ts (hunk 6 above)
Comment (trs) [reply lib/services/dialog/index.ts:207]: For sure, this code is gross and I think deserves a refactor

Comment (theProf) [reply lib/services/dialog/index.ts:207]: for later days

**Comment by trs** [comment]
bors try

**Pull request closed without merging**